Plugins/ExaTrkX/include/Acts/Plugins/ExaTrkX/detail/ConnectedComponents.cuh

```diff
@@ -0,0 +1,173 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Plugins/ExaTrkX/detail/CudaUtils.cuh"
+
+#include <thrust/execution_policy.h>
+#include <thrust/scan.h>
+
+namespace Acts::detail {
+
+template <typename T>
+__device__ void swap(T &a, T &b) {
+  T tmp = a;
+  a = b;
+  b = tmp;
+}
+
+/// Implementation of the FastSV algorithm as shown in
+/// https://arxiv.org/abs/1910.05971
+template <typename TEdge, typename TLabel>
+__global__ void labelConnectedComponents(std::size_t numEdges,
+                                         const TEdge *sourceEdges,
+                                         const TEdge *targetEdges,
+                                         std::size_t numNodes, TLabel *labels,
+                                         TLabel *labelsNext) {
+  // Currently this kernel works only with 1 block
+  assert(gridDim.x == 1);
+
+  for (std::size_t i = threadIdx.x; i < numNodes; i += blockDim.x) {
+    labels[i] = i;
+    labelsNext[i] = i;
+  }
+
+  __shared__ bool changed;
+
+  do {
+    changed = false;
+    __syncthreads();
+
+    //printf("Iteration %i\n", n);
+
+    // Tree hooking for each edge;
+    for (std::size_t i = threadIdx.x; i < numEdges; i += blockDim.x) {
+      auto u = sourceEdges[i];
+      auto v = targetEdges[i];
+
+      if (labels[u] == labels[labels[u]] && labels[v] < labels[u]) {
+        labelsNext[labels[u]] = labels[v];
+        changed = true;
+        //printf("Edge (%i,%i): set labelsNext[%i] = labels[%i] = %i\n", u, v, labels[u], v, labels[v]);
+      } else if (labels[v] == labels[labels[v]] && labels[u] < labels[v]) {
+        labelsNext[labels[v]] = labels[u];
+        changed = true;
+        //printf("Edge (%i,%i): set labelsNext[%i] = labels[%i] = %i\n", u, v, labels[v], u, labels[u]);
+      } else {
+        //printf("Edge (%i,%i): no action\n", u, v);
+      }
+    }
+    __syncthreads();
+
+    for (std::size_t i = threadIdx.x; i < numNodes; i += blockDim.x) {
+      labels[i] = labelsNext[i];
+    }
+
+    /*if(threadIdx.x == 0 ) {
```

Comment:
Don't we need a synchronization point here? :thinking:

Reply:
Hmm my naive view would be that different warps would otherwise interfer here with each other. before we write from `labels -> labelsNext`, and after from `labelsNext -> labels`...